tests/unit/test_job_retry.py

```diff
@@ -240,3 +243,84 @@ def test_raises_on_job_retry_on_result_with_non_retryable_jobs(client):
         ),
     ):
         job.result(job_retry=google.api_core.retry.Retry())
+
+
+def test_query_and_wait_retries_job_for_DDL_queries():
+    """
+    Specific test for retrying DDL queries with "jobRateLimitExceeded" error:
+    https://github.com/googleapis/python-bigquery/issues/1790
+    """
+    freezegun.freeze_time(auto_tick_seconds=50)
+    client = mock.create_autospec(Client)
+    client._call_api.__name__ = "_call_api"
+    client._call_api.__qualname__ = "Client._call_api"
+    client._call_api.__annotations__ = {}
+    client._call_api.__type_params__ = ()
+    client._call_api.side_effect = (
+        {
+            "jobReference": {
+                "projectId": "response-project",
+                "jobId": "abc",
+                "location": "response-location",
+            },
+            "jobComplete": False,
+        },
+        google.api_core.exceptions.InternalServerError(
+            "job_retry me", errors=[{"reason": "jobRateLimitExceeded"}]
+        ),
+        google.api_core.exceptions.BadRequest(
+            "retry me", errors=[{"reason": "jobRateLimitExceeded"}]
+        ),
+        {
+            "jobReference": {
+                "projectId": "response-project",
+                "jobId": "abc",
+                "location": "response-location",
+            },
+            "jobComplete": True,
+            "schema": {
+                "fields": [
+                    {"name": "full_name", "type": "STRING", "mode": "REQUIRED"},
+                    {"name": "age", "type": "INT64", "mode": "NULLABLE"},
+                ],
+            },
+            "rows": [
+                {"f": [{"v": "Whillma Phlyntstone"}, {"v": "27"}]},
+                {"f": [{"v": "Bhetty Rhubble"}, {"v": "28"}]},
+                {"f": [{"v": "Phred Phlyntstone"}, {"v": "32"}]},
+                {"f": [{"v": "Bharney Rhubble"}, {"v": "33"}]},
+            ],
+        },
+    )
+    rows = _job_helpers.query_and_wait(
+        client,
+        query="SELECT 1",
+        location="request-location",
+        project="request-project",
+        job_config=None,
+        page_size=None,
+        max_results=None,
+        retry=google.api_core.retry.Retry(),
+        job_retry=google.api_core.retry.Retry(),
+    )
+    assert len(list(rows)) == 4
+
+    request_path = "/projects/request-project/queries"  # Updated path
+    response_path = "/projects/response-project/jobs/abc"
```

Comment:
Calling `GET` to this path indicates to me that something went very wrong. Somehow the client thought the job has finished.

Reply:
For your reference, here's my thought process:

1. Since the jobs.query says that the query isn't complete, we should end up here: https://github.com/googleapis/python-bigquery/blob/6559dde1568b2d07c1e0a142194f2b370efb8983/google/cloud/bigquery/_job_helpers.py#L479C17-L479C30
2. That calls QueryJob.result(), which I remember calls `jobs.getQueryResults()` to wait for the query to finish and only then calls `jobs.get()` https://github.com/googleapis/python-bigquery/blob/6559dde1568b2d07c1e0a142194f2b370efb8983/google/cloud/bigquery/job/query.py#L1584-L1590 but that appears to be incorrect. You might have identified a bug / regression with this.

Reply:
In the meantime, we can add a TODO pointing to https://github.com/googleapis/python-bigquery/pull/1797 that we need to update this order.